contracts/ServiceNodeContribution.sol

```diff
@@ -0,0 +1,184 @@
+        SENT.approve(address(stakingRewardsContract), stakingRequirement);
+        // TODO sean change the params to pass in all the contributor details also
+        IServiceNodeRewards.Contributor[] memory contributors = new IServiceNodeRewards.Contributor[](1);
+        contributors[0] = IServiceNodeRewards.Contributor(operator, stakingRequirement);
+        stakingRewardsContract.addBLSPublicKey(blsPubkey, blsSignature, serviceNodeParams, contributors);
+
+        emit Finalized(serviceNodeParams.serviceNodePubkey);
+    }
+
+    // TODO rescue funds remaining after finalising
+
+
+    /**
+     * @notice Allows contributors to withdraw their stake before the node is finalized.
+     * @dev Withdrawals are only allowed if the node has not been finalized or cancelled. The operator cannot withdraw their contribution through this method. Operator should call cancelNode() instead.
+     */
+    function withdrawStake() public {
+        require(contributions[msg.sender] > 0, "You have not contributed.");
+        require(!finalized, "Node has already been finalized.");
+        require(msg.sender != operator, "Operator cannot withdraw");
+        uint256 refundAmount = contributions[msg.sender];
+        contributions[msg.sender] = 0;
+        numberContributors -= 1;
+        totalContribution -= refundAmount;
+        SENT.safeTransfer(msg.sender, refundAmount);
+        emit StakeWithdrawn(msg.sender, refundAmount);
+    }
+
+    /**
+     * @notice Cancels the service node setup. Will refund the operator and the contributors will be able to call withdrawStake to get their stake back.
+     * @dev This can only be done by the operator and only if the node has not been finalized or already cancelled.
+     */
+    function cancelNode() public onlyOperator {
+        require(!finalized, "Cannot cancel a finalized node.");
+        require(!cancelled, "Node has already been cancelled.");
+        cancelled = true;
+        uint256 refundAmount = contributions[msg.sender];
+        contributions[msg.sender] = 0;
+        numberContributors -= 1;
+        totalContribution -= refundAmount;
+        SENT.safeTransfer(msg.sender, refundAmount);
+        emit Cancelled(serviceNodeParams.serviceNodePubkey);
+    }
+
+    //////////////////////////////////////////////////////////////
+    //                                                          //
+    //                Non-state-changing functions              //
+    //                                                          //
+    //////////////////////////////////////////////////////////////
+
+    /**
+     * @notice Calculates the minimum contribution amount.
+     * @dev The minimum contribution is dynamically calculated based on the number of contributors and the staking requirement. It returns math.ceilDiv of the calculation
+     * @return The minimum contribution amount.
+     */
+    function minimumContribution() public view returns (uint256) {
+        if (operatorContribution == 0)
```

Comment:
If you set maxContributors to 1, this contract gets stuck because of numberContributors and also this staking requirement override on the first contribution (but also that we lock the initial contribution to 25pct).

Reply:
Also, if we pull this out to a `uint256 public immutable MIN_OPERATOR_CONTRIBUTION` and calculate it in the constructor, it lets people programatically query the value instead of having to hard-code it into their programs, otherwise there's no way in the interface exposed to access this value.